vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+     *      }
+     *
+     *      for (const RRSegmentId& segment : segments()) {
+     *        // Do something with each segment
+     *      }
+     */
+    node_range nodes() const;
+    edge_range edges() const;
+    switch_range switches() const;
+    segment_range segments() const;
+
+    /* Node-level attributes */
+    size_t node_index(const RRNodeId& node) const; /* TODO: deprecate this accessor as outside functions should use RRNodeId */
+
+    /* get the type of a RRGraph node : types of each node, can be channel wires (CHANX or CHANY) or 
+     *                                  logic block pins(OPIN or IPIN) or virtual nodes (SOURCE or SINK)
+     *                                  see t_rr_type definition for more details
+     */
+    t_rr_type node_type(const RRNodeId& node) const;
+
+    /* Get coordinate of a node. (xlow, xhigh, ylow, yhigh):
+     *   For OPIN/IPIN/SOURCE/SINK, xlow = xhigh and ylow = yhigh
+     *   FIXME: should confirm if this is still the case when a logic block has a height > 1
+     *   For CHANX/CHANY, (xlow, ylow) and (xhigh, yhigh) represent
+     *   where the routing segment starts and ends.
+     *   Note that our convention alway keeps
+     *   xlow <= xhigh and ylow <= yhigh
+     *   Therefore, (xlow, ylow) is a starting point for a CHANX/CHANY in INC_DIRECTION
+     *   (xhigh, yhigh) is a starting point for a CHANX/CHANY in INC_DIRECTION
+     *                           
+     *   Example :
+     *   CHANX in INC_DIRECTION  
+     *   (xlow, ylow)                (xhigh, yhigh)
+     *        |                           |
+     *       \|/                         \|/
+     *        ---------------------------->
+     *
+     *   CHANX in DEC_DIRECTION  
+     *   (xlow, ylow)                (xhigh, yhigh)
+     *        |                           |
+     *       \|/                         \|/
+     *        <----------------------------
+     *
+     *   CHANY in INC_DIRECTION
+     *       
+     *      /|\ <-------(xhigh, yhigh)
+     *       |
+     *       |  <-------(xlow, ylow)
+     *
+     *   CHANY in DEC_DIRECTION
+     *       
+     *       |  <-------(xhigh, yhigh)
+     *       |
+     *      \|/ <-------(xlow, ylow)
+     */
+    short node_xlow(const RRNodeId& node) const;
+    short node_ylow(const RRNodeId& node) const;
+    short node_xhigh(const RRNodeId& node) const;
+    short node_yhigh(const RRNodeId& node) const;
+    /* Get the length of a routing track. 
+     * Note that it is ONLY meaningful for CHANX and CHANY, which represents the number of logic blocks that a routing track spans
+     * For nodes that are OPIN/IPIN/SOURCE/SINK, the length is supposed to be always 1
+     */
+    short node_length(const RRNodeId& node) const;
+    /* A short-cut function to get coordinates of a node */
+    vtr::Rect<short> node_bounding_box(const RRNodeId& node) const;
+
+    /* Get node starting and ending points in routing channels. 
```

Comment:
Not clear what "figure for the bounding box" means.
Explain what these functions return for unidirectional architectures, bidirectional architectures, and non-wire nodes.